### PR TITLE
ECER-826: Messages Backend

### PR DESCRIPTION
**Pull request by farzadnadiri**
---
name: Messages Backend
about: Messages required backend logic and APIs added to the project
---

## Title
ECER-{826}: Messages Backend logic added to the project.

## Description

- Communication Repository
- Communication Mapper
- Communication Endpoint
- Communication Tests
- Communication Handler

## Related Jira Issue(s)

- ECER-826
- ECER-825

## Checklist
- [x] I have tested these changes locally.
- [x] Changes to backend endpoints are covered by passing integration tests.
- [ ] I have added or updated the necessary documentation.

### CODE REVIEW

**Comment by pcatkins** [comment]
@farzadnadiri A few small details on PR format:

- can you please make the base branch for this PR `stories/ECER-779` we are trying to use Gitflow workflow when possible.
- also please include the jira ticket number in the title i.e. _ECER-826: Messages Backend_ (apologies if the PR template is unclear — to be improved).

**Comment by farzadnadiri** [comment]
@pcatkins I made the necessary changes. Sorry for the confusion.

File: src/ECER.Tests/Integration/Utilities/DataverseSdk/QueryTests.cs [hunk 1]
```diff
@@ -33,7 +33,6 @@ public void WhereNotIn_Filtered()
     {
         var statuses = new[] { ecer_Application_StatusCode.Draft, ecer_Application_StatusCode.Complete };
         var query = dataverseContext.ecer_ApplicationSet.WhereNotIn(a => a.StatusCode!.Value, statuses).ToList();
-        query.ShouldNotBeEmpty();
```
Comment (ytqsl) [inline src/ECER.Tests/Integration/Utilities/DataverseSdk/QueryTests.cs:36]: I fixed the test in another PR so you should merge master into your branch. In general, this is not a good practice to arbitrarily delete assertions that were there.

File: src/ECER.Tests/Integration/Resources/Accounts/Communications/CommunicationRepositoryTests.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  {
+    // Arrange
+    var communicationId = "9039a260-87eb-4b53-a7b3-0ce5379c4c8b"; // Example ID
+
+    // Act
+    var communications = await repository.Query(new CommunicationQuery { ById = communicationId });
+
+    communications.ShouldBeEmpty();
+    // Assert
+    //communications.ShouldHaveSingleItem();
```
Comment (ytqsl) [inline src/ECER.Tests/Integration/Resources/Accounts/Communications/CommunicationRepositoryTests.cs:38]: Please not commented code unless required. Either delete or finish the assertion.

File: src/ECER.Tests/Integration/RegistryApi/CommunicationTests.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+namespace ECER.Tests.Integration.RegistryApi;
+
+public class CommunicationsTests : RegistryPortalWebAppScenarioBase
+{
+  public CommunicationsTests(ITestOutputHelper output, RegistryPortalWebAppFixture fixture) : base(output, fixture)
+  {
+  }
+
+  [Fact]
+  public async Task GetCommunications_ReturnsCommunications()
```
Comment (ytqsl) [inline src/ECER.Tests/Integration/RegistryApi/CommunicationTests.cs:15]: You probably want to assert that there is returned payload

File: src/ECER.Resources.Accounts/Communications/CommunicationRepositoryMapper.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  public CommunicationRepositoryMapper()
+  {
+    CreateMap<Communication, ecer_Communication>(MemberList.Source)
+              .ForMember(d => d.ecer_CommunicationId, opts => opts.MapFrom(s => s.Id))
+              .ForMember(d => d.ecer_Message, opts => opts.MapFrom(s => s.Body))
+              .ForMember(d => d.ecer_Name, opts => opts.MapFrom(s => s.Subject))
+              .ForMember(d => d.ecer_Acknowledged, opts => opts.MapFrom(s => s.Acknowledged))
+              .ForMember(d => d.ecer_DateNotified, opts => opts.MapFrom(s => s.NotifiedOn))
+              .ForMember(d => d.StatusCode, opts => opts.MapFrom(s => s.Status))
+              .ReverseMap();
```
Comment (ytqsl) [inline src/ECER.Resources.Accounts/Communications/CommunicationRepositoryMapper.cs:18]: This map should be ecer_Communication to Communication only, and no reversal is needed

File: src/ECER.Resources.Accounts/Communications/CommunicationRepositoryMapper.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+    CreateMap<CommunicationStatus, ecer_Communication_StatusCode>()
+            .ConvertUsingEnumMapping(opts => opts.MapByName(true))
```
Comment (ytqsl) [inline src/ECER.Resources.Accounts/Communications/CommunicationRepositoryMapper.cs:21]: Same as above

File: src/ECER.Resources.Accounts/Communications/CommunicationRepository.cs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+  private readonly EcerContext context;
+  private readonly IMapper mapper;
+
+  public CommunicationRepository(EcerContext context, IMapper mapper)
+  {
+    this.context = context;
+    this.mapper = mapper;
+  }
+
+  public async Task<CommunicationsStatus> NotificationStatus(string userId)
```
Comment (ytqsl) [inline src/ECER.Resources.Accounts/Communications/CommunicationRepository.cs:18]: This method needs a rename to something that makes sense. Also, unless the aggregation and count are done in Dataverse (it is possible), there's no value having this query - the manager can call the Query method, roll it up and return the status.

File: src/ECER.Managers.Registry/CommunicationMapper.cs [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+{
+  internal class CommunicationMapper : Profile
+  {
+
+    public CommunicationMapper()
+    {
+      CreateMap<Communication, ECER.Managers.Registry.Contract.Communications.Communication>()
+           .ForMember(dest => dest.Id, opt => opt.MapFrom(src => src.Id ?? string.Empty))
+           .ForMember(dest => dest.Text, opt => opt.MapFrom(src => src.Body))
+           .ReverseMap()
```
Comment (ytqsl) [inline src/ECER.Managers.Registry/CommunicationMapper.cs:14]: No reversal is needed, and the map should be flipped

File: src/ECER.Clients.RegistryPortal/ECER.Clients.RegistryPortal.Server/Communications/CommunicationMapper.cs [hunk 8]
```diff
@@ -0,0 +1,6 @@
+﻿namespace ECER.Clients.RegistryPortal.Server.Communications
+{
+  internal class CommunicationMapper
```
Comment (ytqsl) [inline src/ECER.Clients.RegistryPortal/ECER.Clients.RegistryPortal.Server/Communications/CommunicationMapper.cs:3]: I assume you need to complete this before merging?

File: src/ECER.Clients.RegistryPortal/ECER.Clients.RegistryPortal.Server/Communications/CommunicationsEndpoints.cs [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        Id = i.Id!,
+        Subject = i.Subject,
+        Text = i.Text
+      }));
+    })
+     .WithOpenApi("Handles messages queries", string.Empty, "message_get")
+     .RequireAuthorization();
+
+
+    endpointRouteBuilder.MapGet("/api/messages/status", async (HttpContext httpContext, IMessageBus messageBus) =>
```
Comment (ytqsl) [inline src/ECER.Clients.RegistryPortal/ECER.Clients.RegistryPortal.Server/Communications/CommunicationsEndpoints.cs:31]: Not a very restul uri. I think it is worth considering moving the message status to to the userinfo endpoint - if a user if found, return the aggregated message number.

**Review by ytqsl — CHANGES_REQUESTED** [review]

File: src/ECER.Clients.RegistryPortal/ECER.Clients.RegistryPortal.Server/Communications/CommunicationsEndpoints.cs (hunk 9 above)
Comment (farzadnadiri) [reply src/ECER.Clients.RegistryPortal/ECER.Clients.RegistryPortal.Server/Communications/CommunicationsEndpoints.cs]: we talked about it and decided to keep it as it is for now

File: src/ECER.Managers.Registry/CommunicationMapper.cs (hunk 7 above)
Comment (farzadnadiri) [reply src/ECER.Managers.Registry/CommunicationMapper.cs]: I followed the application as a reference. in the ApplicationHandler we convert Resources to Contract.Applications. In the communications we need that convert as well.

**Review by ytqsl — CHANGES_REQUESTED** [review]
We need to check why the helm chart changes are part of this PR

**Review by ytqsl — CHANGES_REQUESTED** [review]
We need to check why the helm chart changes are part of this PR

File: src/ECER.Clients.RegistryPortal/ECER.Clients.RegistryPortal.Server/Communications/CommunicationsEndpoints.cs [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+public class CommunicationsEndpoints : IRegisterEndpoints
+{
+  public void Register(IEndpointRouteBuilder endpointRouteBuilder)
+  {
+    endpointRouteBuilder.MapGet("/api/messages", async (HttpContext httpContext, IMessageBus messageBus) =>
+    {
+      var userContext = httpContext.User.GetUserContext();
+      var query = new CommunicationsQuery();
+      query.ByIdentity = userContext!.Identity!;
+      query.ByStatus = new CommunicationStatus[] { CommunicationStatus.NotifiedRecipient, CommunicationStatus.Acknowledged };
```
[the remark this answers is not in the document]
Comment (farzadnadiri) [reply src/ECER.Clients.RegistryPortal/ECER.Clients.RegistryPortal.Server/Communications/CommunicationsEndpoints.cs:17]: In Confluence they mentioned that :
Portal should only display communications if NOT Draft.  Allowing staff to compose the message before ready to be shown.

**Review by ytqsl — CHANGES_REQUESTED** [review]

**Review by ytqsl — APPROVED** [review]

**Pull request merged**